File: LC/rough/local_practice/p46.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int m, n;
    int numIslands(vector<vector<char>> &grid)
    {
        int ctr = 0;
        m = grid.size();
        n = grid[0].size();

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (grid[i][j] == '1')
                {
                    ctr++;
                    dfs(grid, i, j);
                }
            }
        }

        return ctr;
    }

    void dfs(vector<vector<char>> &grid, int i, int j)
    {
        if (i < 0 || j < 0 || i >= m || j >= n)
        {
            return;
        }
        if (grid[i][j] == '1')
        {
            grid[i][j] = '2';

            dfs(grid, i + 1, j);
            dfs(grid, i, j + 1);
            dfs(grid, i - 1, j);
            dfs(grid, i, j - 1);
        }

        return;
    }
};

File: LC/rough/local_practice/p46.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        r = len(board)
        c = len(board[0])

        edges = []

        for j in range(c):
            edges.append((0, j))
            edges.append((r - 1, j))
        for i in range(r):
            edges.append((i, 0))
            edges.append((i, c - 1))

        def dfs(i, j):
            board[i][j] = "S"
            dir = [[1, 0], [0, 1], [-1, 0], [0, -1]]

            for dxy in dir:
                nx = dxy[0] + i
                ny = dxy[1] + j

                if nx < 0 or nx >= r or ny < 0 or ny >= c:
                    continue

                if board[nx][ny] == "O":
                    board[nx][ny] = "S"
                    dfs(nx, ny)

        for i in range(r):
            for j in range(c):
                if (i, j) in edges and board[i][j] == "O":
                    dfs(i, j)
        for i in range(r):
            for j in range(c):
                if board[i][j] != "S":
                    board[i][j] = "X"
        for i in range(r):
            for j in range(c):
                if board[i][j] == "S":
                    board[i][j] = "O"
